core/workspace_comment.js

```diff
@@ -0,0 +1,370 @@
+/**
+ * @license
+ * Visual Blocks Editor
+ *
+ * Copyright 2017 Google Inc.
+ * https://developers.google.com/blockly/
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Object representing a code comment on the workspace.
+ * @author fenichel@google.com (Rachel Fenichel)
+ */
+'use strict';
+
+goog.provide('Blockly.WorkspaceComment');
+
+goog.require('Blockly.Events.CommentChange');
+goog.require('Blockly.Events.CommentCreate');
+goog.require('Blockly.Events.CommentDelete');
+goog.require('Blockly.Events.CommentMove');
+
+goog.require('goog.math.Coordinate');
+
+
+/**
+ * Class for a workspace comment.
+ * @param {!Blockly.Workspace} workspace The block's workspace.
+ * @param {string} content The content of this workspace comment.
+ * @param {number} height Height of the comment.
+ * @param {number} width Width of the comment.
+ * @param {string=} opt_id Optional ID.  Use this ID if provided, otherwise
+ *     create a new ID.
+ * @constructor
+ */
+Blockly.WorkspaceComment = function(workspace, content, height, width, opt_id) {
+  /** @type {string} */
+  this.id = (opt_id && !workspace.getCommentById(opt_id)) ?
+      opt_id : Blockly.utils.genUid();
+
+  workspace.addTopComment(this);
+
+  /**
+   * The comment's position in workspace units.  (0, 0) is at the workspace's
+   * origin; scale does not change this value.
+   * @type {!goog.math.Coordinate}
+   * @protected
+   */
+  this.xy_ = new goog.math.Coordinate(0, 0);
+
+  /**
+   * The comment's height in workspace units.  Scale does not change this value.
+   * @type {number}
+   * @private
+   */
+  this.height_ = height;
+
+  /**
+   * The comment's width in workspace units.  Scale does not change this value.
+   * @type {number}
+   * @private
+   */
+  this.width_ = width;
+
+  /**
+   * @type {!Blockly.Workspace}
+   */
+  this.workspace = workspace;
+
+  /**
+   * @protected
+   * @type {boolean}
+   */
+  this.RTL = workspace.RTL;
+
+  /**
+   * @type {boolean}
+   * @private
+   */
+  this.deletable_ = true;
+
+  /**
+   * @type {boolean}
+   * @private
+   */
+  this.movable_ = true;
+
+  /**
+   * @protected
+   * @type {!string}
+   */
+  this.content_ = content;
+
+  /**
+   * @package
+   * @type {boolean}
+   */
+  this.isComment = true;
+
+  Blockly.WorkspaceComment.fireCreateEvent(this);
```

Comment:
Playing around with comments in the vertical_toolbox, I think it would be useful to have the comment create event include the comment's contents in a top level property (instead of having to find it in the xml).

Reply:
Should this also be in a `Blockly.Events.isEnabled()` check like in `dispose` below?

Reply:
Actually I'm not sure this will matter much because we'll probably use the xml for the x, y, width, and height too